### ISSUE TITLE
Timoshenko beams and reinforced concrete

### ISSUE THREAD

**Issue by gauthier12** [mfem/mfem#1666]
Dear all, 
I would like to use MFEM to model reinforced concrete with volumic elements for the concrete and Timoshenko beams for the steel bars.
The steps would be first to implement Timoshenko beams, then implement the snapping method we currently use to link the concrete to the bars.
I've read in issue #342 that it would be possible to implement Timoshenko beams but I have difficulties creating and using a simple beam mesh but in 3 dimensions.
I have attached a sample mesh 
[1D_beam.mesh.txt](https://github.com/mfem/mfem/files/4994218/1D_beam.mesh.txt)

I am able to read the mesh with mfem example 1 and write the output (results are irrelevant for the moment) but I am unable to visualize the mesh with glvis or the paraview output.
With glvis, I get the following error : 
```
MFEM abort: Invalid boundary element type
 ... in function: virtual void VisualizationSceneSolution3d::Prepare()
 ... in file: vssolution3d.cpp:1846

```
and with the paraview output, MFEM is able to write a vtu file but I can't open it with paraview, I get the following error in paraview
```
ERROR: In /builddir/build/BUILD/ParaView-v4.4.0-source/VTK/Filters/Geometry/vtkUnstructuredGridGeometryFilter.cxx, line 1393
vtkUnstructuredGridGeometryFilter (0x4c0f4d0): Cell type UnknownClass(68) is not a 3D cell.
```
I've seen in the mesh file that there is 2 dimensions (dimension and Vdim) but I'am not sure to clearly understand the difference between the two dimensions and I didn't find explanation in the documentation. Is it element dimension and physical space dimension ?

Would you have advice about how to implement 3D Timoshenko beams ?
Woud it be possible to mix, within a mesh, structural elements like beams and volumic elements ?

Cheers, 
Gauthier Folzan

**Comment by v-dobrev** [comment]
Hi @gauthier12,

The short answer for GLVis is that, at the moment, it does not support 1D meshes embedded in 2D or 3D.

One option is to create a function that converts the 1D mesh into 2D mesh by extruding e.g. in +y direction by some small amount. This strip mesh can then be visualized by GLVis. As a starting point, you can use the functions https://github.com/mfem/mfem/blob/b9a40daf5b3b6b5338907c3936941415abd5e4b4/mesh/mesh.cpp#L10639

and https://github.com/mfem/mfem/blob/b9a40daf5b3b6b5338907c3936941415abd5e4b4/fem/gridfunc.cpp#L3728-L3729

These are the functions used in GLVis to create a 2D mesh + solution from 1D mesh (in 1D space) + 1D solution, see here: https://github.com/GLVis/glvis/blob/95aecd8cf3e3038a36366cef14d5d2986a6a8c1a/glvis.cpp#L2145-L2146.

Regarding the Paraview output, the best people to answer your question are probably @bslazarov and @pazner.

Regarding mixing 1D and 3D elements in the same mesh file -- this is currently not possible, but in a longer term, it something that we are considering to add. In the mean time, using two separate meshes (one 1D and second 3D) may be possible, however linking the equations for the two parts may be a bit tricky.

Regarding 3D Timoshenko beams, it should be possible to implement but I'm not familiar with the details apart from some basics. What are the equations you are looking to solve?

**Comment by pazner** [comment]
For what it's worth, I can run example 1 with your mesh and view the results in ParaView. I am saving the solution with a ParaView data collection like this:
```c++
ParaViewDataCollection pd("Example1", &mesh);
pd.SetPrefixPath("ParaView");
pd.RegisterField("solution", &x);
pd.SetLevelsOfDetail(order);
pd.SetCycle(0);
pd.SetTime(0.0);
pd.Save();
```
I have tested that the resulting output works with ParaView 5.7 and 5.8. Perhaps you are using an older version of ParaView.

**Comment by gauthier12** [comment]
@pazner You are right. I was sure I was using the spack installed version of paraview but I was using the old, system one.
Viewing with paraview 5.8 works indeed. Sorry about the mistake and thank you for pointing it out.

@v-dobrev Thank you for your advice and your answers. I've found in the sources the BlockNonlinearForm class which seems to combine two fespaces. I think it may be possible to take inspiration from it to combine the 1D fespace and the 3D fespace and implement an additional integrator to add the coupling terms as if it was a monolithic multiphysic problem.

To begin with, I would start just with a quasi static mechanical equilibrium with elastic behaviour but more complex behaviour laws could be used later.

Cheers
Gauthier

**Comment by v-dobrev** [comment]
Currently, the spaces in `BlockNonlinearForm` are assumed to be on the same mesh. Adding support for different meshes in the trial and test spaces (for integration over a common subset of mesh entities) will require some work. We are definitely interested in such capability but it is currently not on our short term plan. However, I'll be happy to provide assistance, if you want to work on such capability for your project.

**Comment by gauthier12** [comment]
I realized that the `BlockNonlinearForm` assumed that the spaces share the same mesh as I found loops on the first space element list but then looping on all spaces. Nevertheless, even if there some work, knowing that it is a valid start point is a valuable information. Thank you for your advice. I will start from there but if you don't mind, I think I may come back to you with other questions on this subject.
